### PR TITLE
Feature/gsk 1618 run python tests on windows builder

### PR DESCRIPTION
**Pull request by Hartorn**
## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [ ] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [ ] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [ ] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: .github/workflows/test-python-client.yml
Comment (andreybavt) [inline .github/workflows/test-python-client.yml:99]: linting should happen first because it's a lot faster compared to tests execution. In case there are linting issues it'll fail fast

File: python-client/giskard/datasets/base/__init__.py
Comment (andreybavt) [inline python-client/giskard/datasets/base/__init__.py:8]: @Hartorn is it `black` that did the reformatiing?

Comment (andreybavt) [reply python-client/giskard/datasets/base/__init__.py:8]: if yes, could you:
- create a separate branch from `main`
- run black on the same files as you modified in this pr
- create a separate PR with only black modifications
- once this pr is merged merge `main` into `feature/gsk-1618-run-python-tests-on-windows-builder` 

This should only add the relevant changes in the diff of this PR, otherwise it's difficult to review :/

Comment (Hartorn) [reply python-client/giskard/datasets/base/__init__.py:8]: It must be black or isort.
Sorry about that, I will start from main with black & isort formatting, then, I will rebase.

Comment (Hartorn) [reply python-client/giskard/datasets/base/__init__.py:8]: I will also disable my toml linter.

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [inline .github/workflows/build-python-client.yml:26]: why did you leave it commented out?

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [inline .github/workflows/build-python-client.yml:57]: should we activate the caching of pdm installation here?

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [inline .github/workflows/build-python-client.yml:73]: we don't probably need all groups here (`dev` for example) do we? `prod`, `test` and `doc` should be fine

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [inline .github/workflows/build-python-client.yml:86]: I think we can save some time by not building the docs for all python version and architectures since we only use 1 kind of setup on readthedocs

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [inline .github/workflows/build-python-client.yml:83]: don't forget to define `GSK_DISABLE_ANALYTICS` as `true` otherwise we'll have a ton of irrelevant telemetry

File: .github/workflows/sonar.yml
Comment (andreybavt) [inline .github/workflows/sonar.yml:66]: it might be faster to use `gradle-build-action` because it might cache gradle installation itself

File: python-client/giskard/datasets/base/__init__.py
Comment (andreybavt) [inline python-client/giskard/datasets/base/__init__.py:678]: I'd say yes, to make sure there's no case where one of client or run_id are None
WDYT @rabah-khalek

File: python-client/giskard/datasets/base/__init__.py
Comment (andreybavt) [inline python-client/giskard/datasets/base/__init__.py:681]: deletion from `NamedTemporaryFile`'s `__exit__` didn't do the job?

File: python-client/tests/fixtures/amazon_review__binary_classification.py
Comment (andreybavt) [inline python-client/tests/fixtures/amazon_review__binary_classification.py:27]: might as well just be a hardcoded URL, no benefit of using join here...

File: python-client/tests/fixtures/drug_classification__multiclass_classification.py
Comment (andreybavt) [inline python-client/tests/fixtures/drug_classification__multiclass_classification.py:18]: same as the other url: can be hardcoded

**Review by andreybavt — CHANGES_REQUESTED** [review]

File: .github/workflows/build-python-client.yml
Comment (Hartorn) [reply .github/workflows/build-python-client.yml:26]: I was wondering if we should run it for mac os or not.
Same for windows-2022

for python3.11, since I changed the pyproject.toml, no need to run, I can remove it

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [reply .github/workflows/build-python-client.yml:57]: ok, could you create a linear card to set the flag to `true` and reference this card it in a comment so that we don't forget

File: .github/workflows/build-python-client.yml
Comment (Hartorn) [reply .github/workflows/build-python-client.yml:83]: Oh, I thought it was for java backend, my bad.

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [reply .github/workflows/build-python-client.yml:26]: got it, lets not overcomplicate it by adding all the runner types. I type of windows should be fine (I believe 2019 is a more common one these days, right?)
In any case let's clean up the commented out runners

File: .github/workflows/build-python-client.yml
Comment (Hartorn) [reply .github/workflows/build-python-client.yml:86]: Will do

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [reply .github/workflows/build-python-client.yml:83]: it's taken into account by pydantic. The settings are defined in `settings.py` and every setting can be overriden from env variable with a GSK_ prefix (kind of like for Spring)

File: python-client/giskard/datasets/base/__init__.py
Comment (Hartorn) [reply python-client/giskard/datasets/base/__init__.py:681]: I had to put delete=False when opening tempfile.NamedTemporaryFile

When we are doing 
```
with tempfile.NamedTemporaryFile(
            prefix="dataset-", suffix=".csv", delete=False
        ) as f:
```
It boths create AND open the file (with mode w)

But as stated in the doc (https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)
>Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows). 

So for this to work for windows, we have to create a temp file, write it, close, read it and then only delete it.

File: .github/workflows/sonar.yml
Comment (Hartorn) [reply .github/workflows/sonar.yml:66]: I will try that, but after this one I think.
I still have to work on CODEOWNER, gradle cache and some other stuff, but this PR is already a bit too big I think

File: .github/workflows/build-python-client.yml
Comment (andreybavt) [reply .github/workflows/build-python-client.yml:73]: but it can be a good motivation to move the dependencies in the right groups and make sure that people don't make mistakes in the future (for example if you add something to dev and not to test the unit test will fail)

File: .github/workflows/sonar.yml
Comment (andreybavt) [reply .github/workflows/sonar.yml:66]: ok, could you create a separate card please?

File: python-client/giskard/datasets/base/__init__.py
Comment (andreybavt) [reply python-client/giskard/datasets/base/__init__.py:681]: ok, sounds good

File: .github/workflows/sonar.yml
Comment (Hartorn) [reply .github/workflows/sonar.yml:66]: Done ! https://linear.app/giskard/issue/GSK-1717/use-gradle-build-action-instead-of-calling-gradlew

File: .github/workflows/build-python-client.yml
Comment (Hartorn) [reply .github/workflows/build-python-client.yml:73]: I will split that into a separated PR then, if it's ok : https://linear.app/giskard/issue/GSK-1717/use-gradle-build-action-instead-of-calling-gradlew

File: .github/workflows/build-python-client.yml
Comment (Hartorn) [reply .github/workflows/build-python-client.yml]: Done : https://linear.app/giskard/issue/GSK-1712/activate-cache-on-pdm-setup-in-github-action

**Review by andreybavt — APPROVED** [review]

**Pull request merged**
